atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/internalschema/persistence/VersionedInternalSchemaMetadata.java

```diff
@@ -28,11 +32,40 @@
  * Note that additive changes may be made to {@link com.palantir.atlasdb.internalschema.InternalSchemaMetadata}
  * without requiring an increase in the version number; increases are only needed when wire-compatibility would be
  * broken, or it is not possible to assign a sensible default value for events occurring in the past.
+ *
+ * Note: Please be careful when using this object's equals() method, as that compares the exact bytes written
+ * in the payload, and thus might consider semantically equivalent payloads to be different.
  */
 @Value.Immutable
 @JsonSerialize(as = ImmutableVersionedInternalSchemaMetadata.class)
 @JsonDeserialize(as = ImmutableVersionedInternalSchemaMetadata.class)
 public interface VersionedInternalSchemaMetadata {
     int version();
     byte[] payload();
+
+    /**
+     * Returns the underlying {@link InternalSchemaMetadata} for this {@link VersionedInternalSchemaMetadata},
+     * provided we are able to deserialize it. Otherwise returns empty.
+     */
+    @Value.Lazy
+    @JsonIgnore
+    default Optional<InternalSchemaMetadata> underlyingData() {
+        return InternalSchemaMetadataPayloadCodec.tryDecode(this);
+    }
+
+    /**
+     * Returns true iff this {@link VersionedInternalSchemaMetadata} object and the other object are knowably
+     * semantically equivalent - that is, they have the same version, which this product version is able to understand,
```

Comment:
I think it will return true even if you have different versions, but still be able decode it into same object - so if only difference is the way you encrypt it. Is that correct?

Reply:
(Discussed offline - there is actually a version check)